fdb-extensions/src/main/java/com/apple/foundationdb/async/RTree.java

```diff
@@ -0,0 +1,2568 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * An implementation of an R-tree. See <a href="https://en.wikipedia.org/wiki/R-tree">this</a>} link for a general
+ * introduction to R-trees.
+ * <br>
+ * The main use-case for R-trees, while they are normally used for spatial querying of actual objects in N-dimensional
+ * space, is to function as a natural extension of regular B+-tree-based indexes in FDB, but spanning into multiple
+ * dimensions. That allows to answer queries using multiple inequalities which is not possible with 1-D indexes.
+ * <br>
+ * Here a short introduction copied from the explanation at the linked wikipedia page. The key idea of the data structure
+ * is to group nearby objects and represent them with their minimum bounding rectangle in the next higher level of the
+ * tree; the "R" in R-tree is for rectangle. Since all objects lie within this bounding rectangle, a query that does not
+ * intersect the bounding rectangle also cannot intersect any of the contained objects. At the leaf level, each
+ * rectangle describes a single object; at higher levels the aggregation includes an increasing number of objects.
+ * This can also be seen as an increasingly coarse approximation of the data set.
+ * <br>
+ * Similar to the B-tree, the R-tree is also a balanced search tree (so all leaf nodes are at the same depth),
+ * organizes the data in pages/nodes, and is designed for storage on disk. Each node can contain a maximum number of
+ * entries, often denoted as {@code M}. It also guarantees a minimum fill (except for the root node).
+ * <br>
+ * One of the key properties of an R-tree is that the minimum bounding rectangles (MBR) of the children of a node in the
+ * tree may overlap which may cause multiple children to intersect with a query even if that query's mbr is just a
+ * single point. An object is only stored in exactly one leaf node of the tree, however, during a search of the
+ * tree multiple paths may have to exhaustively followed in order go find all matching objects of that query.
+ * <br>
+ * The search performance of the tree is strongly linked to the amount of coverage of a child's mbr as well as the
+ * overlap among children's mbrs at each level of the tree. The key difficulty in constructing a search-efficient tree
+ * is to minimize both coverage and overlap while keeping the tree balanced. Variants of the R-tree such as R+-trees and
```

Comment:
_minimizing coverage_ sounds strange. Although it is clear from the context what the practical goal is, the idea of coverage as a measured thing is ambiguous as to whether it's better to cover more (paint) or to cover more tightly (bedsheet). The thing being unambiguously minimized is, I guess, area or perimeter.

Reply:
I renamed `coverage` to `covered area` which is also the term wikipedia seems to be using.